src/libcore/iter/traits/iterator.rs

```diff
@@ -1690,7 +1690,7 @@ pub trait Iterator {
         Self: Sized, F: FnMut(B, Self::Item) -> R, R: Try<Ok=B>
     {
         let mut accum = init;
-        while let Some(x) = self.next() {
+        for x in self {
```

Comment:
I think this one is a bad idea -- it'll use the `&mut impl Iterator` version of iterator because it will try to move `self`.  And that one is typically less efficient.

(This seems like one of those lints that makes more sense outside the standard library than inside super-core things like try_fold.)

Reply:
You are right, after inspecting the assembly it generated additional `store` instruction.
Disabled lint for this function:
```rust
#[allow(clippy::while_let_on_iterator)] // uses mutable borrow resulting in worse assembly
```